Fixes incorrect strategicpatch behavior on deleting un-existing keys

yue9944882 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

when we delete an unexisting key via strategic patch, we'll find the key will be unexpectedly created somehow. this pull fixes the behavior.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67546

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes unexpected new key occurrence when deleting unexisting key via strategic patch 
```

yue9944882 commented:
/sig api-machinery
/kind bug

yue9944882 commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1329,24 +1329,32 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
-		if !ok {
-			// If it's not in the original document, just take the patch value.
-			original[k] = patchV
-			continue
```
the recursion stops right here. so the further directives would all be ignored.

yue9944882 commented:
/test pull-kubernetes-integration

yue9944882 commented:
/test pull-kubernetes-integration

yue9944882 commented:
/assign @mengqiy

nikhita commented on test/integration/apiserver/patch_test.go:
```diff
@@ -128,6 +131,66 @@ func TestPatchConflicts(t *testing.T) {
+
+	deploymentName := "noxu-deployment"
+	podName := "noxu-pod"
+	containerName := "i-almost-forget-noxu"
```
what does `noxu` mean anyway? I assumed it was a made-up word by @deads2k in the integration tests for apiextensions-apiserver but I see it everywhere now. :)

yue9944882 replied:
<img width="288" alt="screen shot 2018-08-21 at 11 58 16 am" src="https://user-images.githubusercontent.com/12049380/44379656-91a48f80-a539-11e8-9f09-c9cb59bcec0a.png">

lavalamp commented:
/assign

mengqiy commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1329,24 +1329,32 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
+		_, foundInOriginal := original[k]
+		if patchV != nil {
+			if !foundInOriginal {
+				original[k] = reflect.Indirect(reflect.New(reflect.TypeOf(patchV))).Interface()
```
Can you add a comment to explain what do you want to do here?

yue9944882 replied:
sure 😉

mengqiy reviewed (commented):
To cover this, you can add test(s) in `strategicpatch/patch_test.go`. You don't have to add an integration test.

yue9944882 committed:
fix strategic patch

yue9944882 commented:
> you can add test(s) in strategicpatch/patch_test.go. You don't have to add an integration test.

yeah. true. i removed the integration test and added some testcases instead.

mengqiy commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1329,41 +1329,64 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
 			subschema, patchMeta, err2 := schema.LookupPatchMetadataForStruct(k)
 			if err2 != nil {
+				// Handling unknown fields
+				if !foundInOriginal {
```
If `err2` is not caused by an unknown field, I guess doing this blindly is not safe.
Ideally, `LookupPatchMetadataForStruct` should returned a typed error. And you only do this when that type of error is returned.

mengqiy commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go:
```diff
@@ -1329,41 +1329,64 @@ func mergeMap(original, patch map[string]interface{}, schema LookupPatchMeta, me
 			subschema, patchMeta, err2 := schema.LookupPatchMetadataForSlice(k)
 			if err2 != nil {
+				// Handling unknown fields
+				if !foundInOriginal {
```
Same comment here.

mengqiy commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go:
```diff
@@ -428,6 +428,44 @@ testCases:
+            $patch: delete
+    modified:
+      retainKeysMap: {}
+  - description: retainKeys map can replace an unexisting merging list
```
This seems to be an breaking behavior change for `replace` directive.
IMHO there is no bug for `replace`. And this behavior change for `replace` seems wrong to me.
#67546 is for `delete` directive.

yue9944882 commented:
/remove-lifecycle stale

yue9944882 commented:
/remove-lifecycle rotten

yue9944882 commented:
/retest

yue9944882 commented:
/test pull-kubernetes-integration
